sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -20,339 +20,13 @@ package org.apache.spark.sql.catalyst.optimizer
-                  val factCol = if (factTable.outputSet.contains(lhs)) lhs else rhs
-                  hasStatistics(dimCol, dimTable) && hasStatistics(factCol, factTable)
-                case _ => false
-              }
-            }
-
-            if (!areStatsAvailable) {
-              emptyStarJoinPlan
-            } else {
-              // Find the subset of dimension tables. A dimension table is assumed to be in a
-              // RI relationship with the fact table. Only consider equi-joins
-              // between a fact and a dimension table to avoid expanding joins.
-              val eligibleDimPlans = allFactJoins.filter { dimTable =>
-                allFactJoinCond.exists {
-                  case cond @ Equality(lhs: AttributeReference, rhs: AttributeReference) =>
-                    val dimCol = if (dimTable.outputSet.contains(lhs)) lhs else rhs
-                    isUnique(dimCol, dimTable)
-                  case _ => false
-                }
-              }
-
-              if (eligibleDimPlans.isEmpty) {
-                // An eligible star join was not found because the join is not
-                // an RI join, or the star join is an expanding join.
-                emptyStarJoinPlan
-              } else {
-                Seq(factTable +: eligibleDimPlans)
-              }
-            }
-        }
-      }
-    }
-  }
-
-  /**
-   * Reorders a star join based on heuristics:
-   *   1) Finds the star join with the largest fact table and places it on the driving
-   *      arm of the left-deep tree. This plan avoids large table access on the inner, and
-   *      thus favor hash joins.
-   *   2) Applies the most selective dimensions early in the plan to reduce the amount of
-   *      data flow.
-   */
-  def reorderStarJoins(
-      input: Seq[(LogicalPlan, InnerLike)],
-      conditions: Seq[Expression]): Seq[(LogicalPlan, InnerLike)] = {
-    assert(input.size >= 2)
-
-    val emptyStarJoinPlan = Seq.empty[(LogicalPlan, InnerLike)]
-
-    // Find the eligible star plans. Currently, it only returns
-    // the star join with the largest fact table.
-    val eligibleJoins = input.collect{ case (plan, Inner) => plan }
-    val starPlans = findStarJoins(eligibleJoins, conditions)
-
-    if (starPlans.isEmpty) {
-      emptyStarJoinPlan
-    } else {
-      val starPlan = starPlans.head
-      val (factTable, dimTables) = (starPlan.head, starPlan.tail)
-
-      // Only consider selective joins. This case is detected by observing local predicates
-      // on the dimension tables. In a star schema relationship, the join between the fact and the
-      // dimension table is a FK-PK join. Heuristically, a selective dimension may reduce
-      // the result of a join.
-      // Also, conservatively assume that a fact table is joined with more than one dimension.
-      if (dimTables.size >= 2 && isSelectiveStarJoin(dimTables, conditions)) {
```

Comment:
@gatorsmile I moved the condition from ```reorderStarJoin``` to ```findStarJoin``` to be used by both ReorderJoin and CostBasedReorder.